tests/Doctrine/Tests/ORM/Persisters/BasicEntityPersisterTypeValueSqlTest.php

```diff
@@ -149,6 +149,11 @@ public function testCountCondition(): void
         $criteria  = new Criteria(Criteria::expr()->eq('value', 'bar'));
         $statement = $persister->getCountSQL($criteria);
         self::assertEquals('SELECT COUNT(*) FROM "not-a-simple-entity" t0 WHERE t0."simple-entity-value" = ?', $statement);
+
+        // Using a criteria object w/ Limit
+        $criteria  = new Criteria(Criteria::expr()->eq('value', 'bar'), null, null, 5000);
+        $statement = $persister->getCountSQL($criteria);
+        self::assertEquals('SELECT COUNT(*) FROM "not-a-simple-entity" t0 WHERE t0."simple-entity-value" = ? LIMIT 5000', $statement);
```

Comment:
Wait a minute… what will this return if there are 5001 elements? 5001, right? If yes, what's the point?

Reply:
Yes it'll return 5001. I'll have to test it better tomorrow, but from my understanding - if the user passes a criteria with a limit to `matching`, the returned collection won't contain more items than specified by the limit. But the count doesn't use the limit, so I think the below scenario is entirely possible:
```php
$c = new Criteria(null, null, null, 5000);
$results = $repository->matching($c);
echo $results->count(); // 2000000
```

Reply:
Ok, then what's the point?

Reply:
Of adding the Limit? Performance. I have a default limit of 10k records for a query, which should be fast, but checking the size of the collection (using `isEmpty`) triggers ` SELECT COUNT(*)` on the entire table. But I would also expect the collection, which is a result of the criteria, to have no more records in it than the maxResults set in the criteria (if it is set). It if isn't set, then yeah the user is asking for a count of the whole table.

Reply:
You say "`SELECT COUNT(*)` on the entire table" But `SELECT COUNT(*) LIMIT 1` is also on the entire table, and the `LIMIT 1` part seems useless here. I can't think how the `LIMIT 1` would improve performance here, because it just counts the entire table, then returns a row containing the count, which will always be just the one row.

Reply:
Ohhh, right, yes. Sorry for being dense, I haven't gotten enough sleep the past few days.
Yeah, to be performant it would have to use a subquery
```sql
SELECT COUNT(*) FROM (SELECT * FROM table WHERE ... LIMIT 1) as f
```
I could refactor it to do that - do you think that's a good idea?

Reply:
I don't, think it's a good idea either, sorry. I mean, it might be performant, but it is also useless as soon as the limit kicks in.